bat.d/41-framework

```diff
@@ -0,0 +1,536 @@
+    #       128=no kernel support/
+    #       254=laptop not supported
+    #
+    # 2. determine method for
+    #    reading battery data                   --> retval $_bm_read,
+    #    reading/writing charging thresholds    --> retval $_bm_thresh,
+    #    reading/writing force discharge        --> retval $_bm_dischg:
+    #       none/natacpi
+    #
+    # 3. define sysfile basenames for natacpi
+    #    stop threshold                         --> retval $_bn_stop,
+    #
+    # 4. determine present batteries
+    #    list of batteries (space separated)    --> retval $_batteries;
+    #
+    # 5. define charge threshold defaults
+    #    stop threshold                         --> retval $_bt_def_stop;
+
+    _batdrv_plugin="framework"
+    _batdrv_kmod="framework_laptop" # kernel module for natacpi
+
+    # check plugin simulation override and denylist
+    if [ -n "$X_BAT_PLUGIN_SIMULATE" ]; then
+        if [ "$X_BAT_PLUGIN_SIMULATE" = "$_batdrv_plugin" ]; then
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.simulate"
+        else
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.simulate_skip"
+            return 1
+        fi
+    elif wordinlist "$_batdrv_plugin" "$X_BAT_PLUGIN_DENYLIST"; then
+        echo_debug "bat" "batdrv_init.${_batdrv_plugin}.denylist"
+        return 1
+    else
+        # check if hardware matches
+        if ! batdrv_is_framework; then
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.no_match"
+            return 1
+        fi
+    fi
+
+    # presume no features at all
+    _natacpi=128
+    # shellcheck disable=SC2034
+    _bm_read="natacpi"
+    _bm_thresh="none"
+    # shellcheck disable=SC2034
+    _bm_dischg="none"
+    _bn_stop=""
+    _batteries=""
+	# framework doesn't support charge start threshold
+    _bt_def_stop=60
+
+    # iterate batteries and check for native kernel ACPI
+    local bd bs
+    local done=0
+    for bd in "$ACPIBATDIR"/BAT[01]; do
+        if [ "$(read_sysf "$bd/present")" = "1" ]; then
+            # record detected batteries and directories
+            bs=${bd##/*/}
+            if [ -n "$_batteries" ]; then
+                _batteries="$_batteries $bs"
+            else
+                _batteries="$bs"
+            fi
+            # skip natacpi detection for 2nd and subsequent batteries
+            [ $done -eq 1 ] && continue
+
+            done=1
+            if [ "$NATACPI_ENABLE" = "0" ]; then
+                # natacpi disabled in configuration --> skip actual detection
+                _natacpi=32
+                continue
+            fi
+
+            if [ -f "$bd/charge_control_end_threshold" ]; then
+                # threshold sysfiles exist
+                _bn_stop="charge_control_end_threshold"
+                _natacpi=254
+            else
+                # nothing detected
+                _natacpi=254
+                continue
+            fi
+
+            if readable_sysf "$bd/$_bn_stop"; then
```

Comment:
you may as well join everything like this: https://github.com/linrunner/TLP/blob/main/bat.d/15-asus#L95